## **Apuntes**

### **Apuntes #1** - Carga Incremental para el `Método 2`

In [ ]:
results_final_df.select('race_id').distinct().collect()

Out[44]: [Row(race_id=148),
 Row(race_id=463),
 Row(race_id=471),
 Row(race_id=496),
 Row(race_id=243),
 Row(race_id=392),
 Row(race_id=540),
 Row(race_id=31),
 Row(race_id=516),
 Row(race_id=85),
 Row(race_id=137),
 Row(race_id=251),
 Row(race_id=451),
 Row(race_id=580),
 Row(race_id=65),
 Row(race_id=458),
 Row(race_id=53),
 Row(race_id=255),
 Row(race_id=481),
 Row(race_id=588),
 Row(race_id=133),
 Row(race_id=296),
 Row(race_id=472),
 Row(race_id=78),
 Row(race_id=322),
 Row(race_id=513),
 Row(race_id=321),
 Row(race_id=362),
 Row(race_id=375),
 Row(race_id=593),
 Row(race_id=108),
 Row(race_id=155),
 Row(race_id=34),
 Row(race_id=193),
 Row(race_id=211),
 Row(race_id=368),
 Row(race_id=530),
 Row(race_id=101),
 Row(race_id=115),
 Row(race_id=126),
 Row(race_id=81),
 Row(race_id=385),
 Row(race_id=28),
 Row(race_id=183),
 Row(race_id=210),
 Row(race_id=436),
 Row(race_id=497),
 Row(race_id=300),
 Row(race_id=406),
 Row(race_id=412),
 Row(race_id=587),
 Row(race_id=76),
 Row(race_id

In [ ]:
# Obtener y mostrar los IDs de carrera únicos
race_id_list = results_final_df.select('race_id').distinct().collect()
for race in race_id_list:
    print(race) 

Row(race_id=148)
Row(race_id=463)
Row(race_id=471)
Row(race_id=496)
Row(race_id=243)
Row(race_id=392)
Row(race_id=540)
Row(race_id=31)
Row(race_id=516)
Row(race_id=85)
Row(race_id=137)
Row(race_id=251)
Row(race_id=451)
Row(race_id=580)
Row(race_id=65)
Row(race_id=458)
Row(race_id=53)
Row(race_id=255)
Row(race_id=481)
Row(race_id=588)
Row(race_id=133)
Row(race_id=296)
Row(race_id=472)
Row(race_id=78)
Row(race_id=322)
Row(race_id=513)
Row(race_id=321)
Row(race_id=362)
Row(race_id=375)
Row(race_id=593)
Row(race_id=108)
Row(race_id=155)
Row(race_id=34)
Row(race_id=193)
Row(race_id=211)
Row(race_id=368)
Row(race_id=530)
Row(race_id=101)
Row(race_id=115)
Row(race_id=126)
Row(race_id=81)
Row(race_id=385)
Row(race_id=28)
Row(race_id=183)
Row(race_id=210)
Row(race_id=436)
Row(race_id=497)
Row(race_id=300)
Row(race_id=406)
Row(race_id=412)
Row(race_id=587)
Row(race_id=76)
Row(race_id=26)
Row(race_id=27)
Row(race_id=332)
Row(race_id=501)
Row(race_id=577)
Row(race_id=384)
Row(race_id=44)
Row(race_

In [ ]:
# Obtener y mostrar los IDs de carrera únicos
race_id_list = results_final_df.select('race_id').distinct().collect()
for race in race_id_list:
    print(race.race_id) 

148
463
471
496
243
392
540
31
516
85
137
251
451
580
65
458
53
255
481
588
133
296
472
78
322
513
321
362
375
593
108
155
34
193
211
368
530
101
115
126
81
385
28
183
210
436
497
300
406
412
587
76
26
27
332
501
577
384
44
159
192
271
253
103
236
329
460
336
12
223
388
417
548
91
409
578
333
222
285
22
128
209
330
372
122
230
319
493
93
157
190
225
232
233
246
360
367
476
539
111
224
519
47
140
177
416
556
132
152
444
185
305
291
325
386
435
473
581
146
259
512
206
1
52
297
363
212
274
182
442
218
280
13
440
86
6
16
474
168
205
572
142
178
3
308
328
20
40
164
429
470
169
500
139
283
359
479
295
431
94
250
377
402
432
582
57
570
292
306
54
120
559
491
96
235
545
452
48
266
507
544
163
191
268
5
258
279
19
397
430
92
419
542
553
560
415
257
299
64
117
227
579
504
554
41
154
309
425
447
455
484
43
15
262
511
112
269
327
563
165
358
366
179
189
207
489
590
37
383
61
127
316
461
502
523
88
197
486
531
318
462
547
107
202
263
374
564
231
9
552
324
586
72
286
398
505
538
175
387
423
381
498


In [ ]:
n = 0
race_id_list = results_final_df.select('race_id').distinct().collect()
for race in race_id_list:
  n = n + 1
print(n)

1035


El siguiente bucle **FOR** realiza lo siguiente:

1. Si la tabla **f1_processed.results** existe elimina todas las **PARTICIONES** que se hayan creado para el archivo ingestado
2. Si no hubiesemos utilizado la sentencia **IF** y la tabla **f1_processed.results** no hubiese existido, nos hubiese devuelto un **ERROR**. Es por eso que utilizamos el **IF**
3. Como estamos ingestando el archivo **results.json** del directorio **2021-03-21**, existen 1035 particiones
4. Con este método, podemos EJECUTAR EL NOTEBOOK CUANTAS VECES QUERAMOS Y NO SE DUPLICARAN LOS DATOS
5. **ALTER TABLE 'tabla' DROP PARTITION** es una sentencia de HIVE
6. **spark._jsparkSession.catalog()**: Accede al catálogo del SparkSession

In [ ]:
race_id_list = results_final_df.select('race_id').distinct().collect()
for race in race_id_list:
  if (spark._jsparkSession.catalog().tableExists("f1_processed.results")):
    spark.sql(f"ALTER TABLE f1_processed.results DROP IF EXISTS PARTITION (race_id = {race.race_id})")

In [ ]:
# Escribimos el archivo con formato PARQUET en la base de datos "f1_processed" y en la tabla "results"
# Antes de ejecutar el código, debemos eliminar el directorio y archivo parquet creado previamente. Esto 
# porque al crear la tabla creará un directorio y archivo, y además, creará la tabla
results_final_df.write.mode("append").partitionBy('race_id').format("parquet").saveAsTable("f1_processed.results")

In [ ]:
spark.read.parquet("/mnt/formula1dl/processed/results").show(truncate=False)

+---------+---------+--------------+------+----+--------+-------------+--------------+------+----+----------+------------+-----------+----+----------------+-----------------+-----------+----------+-----------------------+-------+
|result_id|driver_id|constructor_id|number|grid|position|position_text|position_order|points|laps|time      |milliseconds|fastest_lap|rank|fastest_lap_time|fastest_lap_speed|data_source|file_date |ingestion_date         |race_id|
+---------+---------+--------------+------+----+--------+-------------+--------------+------+----+----------+------------+-----------+----+----------------+-----------------+-----------+----------+-----------------------+-------+
|19232    |657      |113           |14    |19  |1       |1            |1             |8.0   |200 |3:49:17.27|13757270    |null       |null|\N              |null             |Ergast     |2021-03-21|2023-06-15 00:36:26.176|800    |
|19233    |525      |114           |9     |3   |2       |2            |2        

In [ ]:
%sql
-- Vemos que tenemos la data hasta la race_id = 1047
SELECT race_id, COUNT(1)
FROM f1_processed.results
GROUP BY race_id
ORDER BY race_id DESC

race_id count(1) 1047 20 1046 20 1045 20 1044 20 1043 20 1042 20 1041 20 1040 20 1039 20 1038 20 1037 20 1036 20 1035 20 1034 20 1033 20 1032 20 1031 20 1030 20 1029 20 1028 20 1027 20 1026 20 1025 20 1024 20 1023 20 1022 20 1021 20 1020 20 1019 20 1018 20 1017 20 1016 20 1015 20 1014 20 1013 20 1012 20 1011 20 1010 20 1009 20 1008 20 1007 20 1006 20 1005 20 1004 20 1003 20 1002 20 1001 20 1000 20 999 20 998 20 997 20 996 20 995 20 994 20 993 20 992 20 991 20 990 20 989 20 988 20 987 20 986 20 985 20 984 20 983 20 982 20 981 20 980 20 979 20 978 20 977 20 976 20 975 20 974 20 973 20 972 20 971 20 970 20 969 20 968 22 967 22 966 22 965 22 964 22 963 22 962 22 961 22 960 22 959 22 958 22 957 22 956 22 955 22 954 22 953 22 952 22 951 22 950 22 949 22 948 22 945 20 944 20 943 20 942 20 941 20 940 20 939 20 938 20 937 20 936 20 934 20 933 20 932 20 931 20 930 20 929 20 928 20 927 20 926 18 918 20 917 18 916 18 915 21 914 22 913 22 912 22 911 22 910 22 909 22 908 22 907 22 906 22 905 22 904 22 903 22 902 22 901 22 900 22 899 22 898 22 897 22 896 22 895 22 894 22 893 22 892 22 891 22 890 22 888 22 887 22 886 22 885 22 884 22 883 22 882 22 881 22 880 22 879 24 878 24 877 24 876 24 875 24 874 24 873 24 872 24 871 24 870 24 869 24 868 24 867 24 866 24 865 24 864 24 863 24 862 24 861 24 860 24 859 24 858 24 857 24 856 24 855 24 854 24 853 24 852 24 851 24 850 24 849 24 848 24 847 24 846 24 845 24 844 24 843 24 842 24 841 24 839 29 838 20 837 14 836 18 835 35 834 21 833 23 832 20 831 23 830 22 829 20 828 26 827 13 826 34 825 21 824 35 823 19 822 30 821 32 820 22 819 22 818 33 817 22 816 31 815 22 814 36 813 27 812 25 811 21 810 20 809 47 808 17 807 23 806 22 805 17 804 24 803 34 802 22 801 15 800 55 799 18 798 23 797 28 796 16 795 17 794 35 793 26 792 35 791 29 790 22 789 30 788 22 787 16 786 34 785 21 784 16 783 20 782 16 781 24 780 22 779 16 778 33 777 23 776 17 775 25 774 22 773 15 772 26 771 20 770 22 769 19 768 33 767 18 766 31 765 10 764 19 763 21 762 16 761 15 760 30 759 22 758 15 757 33 756 24 755 23 754 17 753 15 752 25 751 23 750 19 749 21 748 33 747 24 746 23 745 21 744 39 743 33 742 32 741 31 740 26 739 17 738 21 737 21 736 20 735 31 734 31 733 26 732 22 731 26 730 24 729 22 728 23 727 25 726 24 725 33 724 29 723 26 722 24 721 21 720 24 719 25 718 19 717 20 716 24 715 20 714 23 713 25 712 17 711 18 710 18 709 23 708 19 707 18 706 23 705 21 704 17 703 23 702 17 701 19 700 17 699 31 698 18 697 19 696 22 695 19 694 17 693 20 692 17 691 18 690 16 689 19 688 18 687 18 686 18 685 24 684 21 683 15 682 18 681 17 680 19 679 18 678 21 677 20 676 21 675 22 674 20 673 20 672 17 671 19 670 18 669 18 668 14 667 23 666 17 665 18 664 20 663 15 662 14 661 17 660 13 659 15 658 16 657 14 656 18 655 18 654 27 653 20 652 27 651 24 650 25 649 23 648 23 647 24 646 17 645 21 644 22 643 23 642 29 641 26 640 24 639 22 638 23 637 24 636 23 635 23 634 22 633 22 632 25 631 31 630 24 629 27 628 25 627 27 626 26 625 25 624 25 623 25 622 26 621 26 620 22 619 28 618 26 617 25 616 23 615 22 614 24 613 29 612 25 611 21 610 25 609 23 608 22 607 25 606 20 605 19 604 30 603 30 602 31 601 31 600 32 599 34 598 30 597 27 596 27 595 28 594 32 593 27 592 27 591 25 590 25 589 24 588 28 587 30 586 26 585 28 584 26 583 25 582 26 581 24 580 26 579 26 578 28 577 23 576 23 575 26 574 27 573 27 572 29 571 27 570 25 569 28 568 30 567 30 566 27 565 25 564 29 563 30 562 27 561 25 560 22 559 23 558 27 557 27 556 34 555 34 554 30 553 29 552 36 551 30 550 31 549 31 548 26 547 31 546 22 545 23 544 22 543 21 542 28 541 27 540 32 539 33 538 31 537 30 536 30 535 29 534 27 533 29 532 30 531 30 530 30 529 30 528 28 527 27 526 30 525 30 524 28 523 26 522 26 521 26 520 26 519 27 518 25 517 28 516 27 515 26 514 26 513 26 512 26 511 27 510 28 509 28 508 29 507 26 506 26 505 27 504 27 503 27 502 27 501 27 500 28 499 28 498 28 497 30 496 30 495 30 494 30 493 28 492 30 491 30 490 29 489 30 488 31 487 31 486 32 485 29 484 30 483 29 482 30 481 30 480 29 479 29 478 29 477 29 476 30 475 31 474 29 4

### **Apuntes #2** - Carga Incremental para el `Método 3`

In [ ]:
results_final_df.select('race_id').distinct().collect()

Out[45]: [Row(race_id=148),
 Row(race_id=463),
 Row(race_id=471),
 Row(race_id=496),
 Row(race_id=243),
 Row(race_id=392),
 Row(race_id=540),
 Row(race_id=31),
 Row(race_id=516),
 Row(race_id=85),
 Row(race_id=137),
 Row(race_id=251),
 Row(race_id=451),
 Row(race_id=580),
 Row(race_id=65),
 Row(race_id=458),
 Row(race_id=53),
 Row(race_id=255),
 Row(race_id=481),
 Row(race_id=588),
 Row(race_id=133),
 Row(race_id=296),
 Row(race_id=472),
 Row(race_id=78),
 Row(race_id=322),
 Row(race_id=513),
 Row(race_id=321),
 Row(race_id=362),
 Row(race_id=375),
 Row(race_id=593),
 Row(race_id=108),
 Row(race_id=155),
 Row(race_id=34),
 Row(race_id=193),
 Row(race_id=211),
 Row(race_id=368),
 Row(race_id=530),
 Row(race_id=101),
 Row(race_id=115),
 Row(race_id=126),
 Row(race_id=81),
 Row(race_id=385),
 Row(race_id=28),
 Row(race_id=183),
 Row(race_id=210),
 Row(race_id=436),
 Row(race_id=497),
 Row(race_id=300),
 Row(race_id=406),
 Row(race_id=412),
 Row(race_id=587),
 Row(race_id=76),
 Row(race_id

In [ ]:
for race_id_list in results_final_df.select('race_id').distinct().collect():
  print(race_id_list)

Row(race_id=148)
Row(race_id=463)
Row(race_id=471)
Row(race_id=496)
Row(race_id=243)
Row(race_id=392)
Row(race_id=540)
Row(race_id=31)
Row(race_id=516)
Row(race_id=85)
Row(race_id=137)
Row(race_id=251)
Row(race_id=451)
Row(race_id=580)
Row(race_id=65)
Row(race_id=458)
Row(race_id=53)
Row(race_id=255)
Row(race_id=481)
Row(race_id=588)
Row(race_id=133)
Row(race_id=296)
Row(race_id=472)
Row(race_id=78)
Row(race_id=322)
Row(race_id=513)
Row(race_id=321)
Row(race_id=362)
Row(race_id=375)
Row(race_id=593)
Row(race_id=108)
Row(race_id=155)
Row(race_id=34)
Row(race_id=193)
Row(race_id=211)
Row(race_id=368)
Row(race_id=530)
Row(race_id=101)
Row(race_id=115)
Row(race_id=126)
Row(race_id=81)
Row(race_id=385)
Row(race_id=28)
Row(race_id=183)
Row(race_id=210)
Row(race_id=436)
Row(race_id=497)
Row(race_id=300)
Row(race_id=406)
Row(race_id=412)
Row(race_id=587)
Row(race_id=76)
Row(race_id=26)
Row(race_id=27)
Row(race_id=332)
Row(race_id=501)
Row(race_id=577)
Row(race_id=384)
Row(race_id=44)
Row(race_

In [ ]:
for race_id_list in results_final_df.select('race_id').distinct().collect():
  print(race_id_list.race_id)

148
463
471
496
243
392
540
31
516
85
137
251
451
580
65
458
53
255
481
588
133
296
472
78
322
513
321
362
375
593
108
155
34
193
211
368
530
101
115
126
81
385
28
183
210
436
497
300
406
412
587
76
26
27
332
501
577
384
44
159
192
271
253
103
236
329
460
336
12
223
388
417
548
91
409
578
333
222
285
22
128
209
330
372
122
230
319
493
93
157
190
225
232
233
246
360
367
476
539
111
224
519
47
140
177
416
556
132
152
444
185
305
291
325
386
435
473
581
146
259
512
206
1
52
297
363
212
274
182
442
218
280
13
440
86
6
16
474
168
205
572
142
178
3
308
328
20
40
164
429
470
169
500
139
283
359
479
295
431
94
250
377
402
432
582
57
570
292
306
54
120
559
491
96
235
545
452
48
266
507
544
163
191
268
5
258
279
19
397
430
92
419
542
553
560
415
257
299
64
117
227
579
504
554
41
154
309
425
447
455
484
43
15
262
511
112
269
327
563
165
358
366
179
189
207
489
590
37
383
61
127
316
461
502
523
88
197
486
531
318
462
547
107
202
263
374
564
231
9
552
324
586
72
286
398
505
538
175
387
423
381
498


In [ ]:
n = 0
for race_id_list in results_final_df.select('race_id').distinct().collect():
  n = n + 1
print(n)

1035


1. Si no hubiesemos utilizado la sentencia **IF** y la tabla **f1_processed.results** no hubiese existido, nos hubiese devuelto un **ERROR**. Es por eso que utilizamos el **IF**
2. **spark.conf.set("spark.sql.sources.partitionOverwriteMode","dynamic")** nos permite utilizar la función **insertInto** en modo **overwrite** y asi **`SOLO SOBRESCRIBIR LAS PARTICIONES QUE HAN SIDO MODIFICADAS`**.
3. Por defecto, si no utilizamos el modo **dynamic** spark utiliza el modo **static**. Esto significa que, al usar operaciones de sobrescritura (como `insertInto()`), **`SE SOBRESCRIBEN TODAS LAS PARTICIONES DEL DESTINO`**, independientemente de si se especifican o no en el DataFrame que estás escribiendo.
4. La configuración **spark.conf.set("spark.sql.sources.partitionOverwriteMode", "dynamic")** sirve para:

    - Sobrescribir solo ciertas particiones: En lugar de reescribir todo el conjunto de datos, solo se actualizan las particiones que cambian.
    - Ahorro de tiempo y recursos: Reduce el tiempo de escritura y el uso de espacio en disco.
    - Facilita actualizaciones: Ideal para actualizar datos de forma incremental, como en datos que llegan regularmente.
5. Por lo tanto, si la tabla ya existe, solo sobrescribirá aquellas particiones modificadas y si existen registros nuevos para dicha partición los agregará. En caso contrario, no hará nada
6. Si no existe la tabla, la creará
7. Con este método, podemos EJECUTAR EL NOTEBOOK CUANTAS VECES QUERAMOS Y NO SE DUPLICARAN LOS DATOS
8. **spark._jsparkSession.catalog()**: Accede al catálogo del SparkSession

In [ ]:
spark.conf.set("spark.sql.sources.partitionOverwriteMode","dynamic")

Vamos a explicar lo que hace esta propiedad. Tenemos una estructura similar para el archivo `results`:

```text
/raw/
├── 2024-10-10/
│   └── results/
│       ├── race_id=1/
│       │   └── part-0000-xxxx.snappy.parquet
│       ├── race_id=2/
│       │   └── part-0000-xxxx.snappy.parquet
│       ├── race_id=3/
│       │   └── part-0000-xxxx.snappy.parquet
│       ├── race_id=4/
│       │   └── part-0000-xxxx.snappy.parquet
│       └── race_id=5/
│           └── part-0000-xxxx.snappy.parquet
└── 2024-10-11/
    └── results/
        ├── race_id=1/
        │   └── part-0000-xxxx.snappy.parquet
        ├── race_id=2/
        │   └── part-0000-xxxx.snappy.parquet
        ├── race_id=3/
        │   └── part-0000-xxxx.snappy.parquet
        ├── race_id=4/
        │   └── part-0000-xxxx.snappy.parquet
        └── race_id=5/
            └── part-0000-xxxx.snappy.parquet
```

Supongamos que en la partición `race_id`, tienes los siguientes registros:

- **Antes de la escritura**:
  - `race_id = 1`: resultado viejo 1
  - `race_id = 2`: resultado viejo 2
  - `race_id = 3`: resultado viejo 3
  - `race_id = 4`: resultado viejo 4
  - `race_id = 5`: resultado viejo 5
  - `race_id = 6`: resultado nuevo

> Usando **Dynamic**:

- **Si cargas nuevos datos solo para `race_id = 3`**:
  - **Después de la escritura**:
    - `race_id = 1`: resultado viejo 1 (se mantiene)
    - `race_id = 2`: resultado viejo 2 (se mantiene)
    - `race_id = 3`: nuevo resultado 3 (se elimina y se escribe nuevamente)
    - `race_id = 4`: resultado viejo 4 (se mantiene)
    - `race_id = 5`: resultado viejo 5 (se mantiene)
    - `race_id = 6`: resultado nuevo (se escribe) 

Esto quiere decir que si cargamos nuevamente el directorio `2024-10-10` y el archivo **results.csv** viene con datos nuevos para el race_id = 3, solo esta partición se actualizará.

> Usando **Static**:

- **Si cargas nuevos datos para `race_id = 3`**:
  - **Después de la escritura**:
    - `race_id = 1`: (se elimina y se escribe nuevamente)
    - `race_id = 2`: (se elimina y se escribe nuevamente)
    - `race_id = 3`: nuevo resultado 3 (se elimina y se escribe nuevamente)
    - `race_id = 4`: (se elimina y se escribe nuevamente)
    - `race_id = 5`: (se elimina y se escribe nuevamente)
    - `race_id = 6`: resultado nuevo(se escribe)

Esto quiere decir que si cargamos nuevamente el directorio `2024-10-10` y el archivo **results.csv** viene con datos nuevos para el race_id = 3, se eliminarán todas las particiones y solo esta partición se actualizará.    

> Resumen Final

- **Dynamic**: Solo sobrescribe el dato para `race_id = 3` y los otros `race_id` permanecen intactos.
- **Static**: Elimina todos los registros y los escribe nuevamente, tambien el nuevo dato para `race_id = 3`.

In [ ]:
# Spark toma el campo particionado y lo deja al último en orden, cada vez que se particiona y se escribe un dataframe.
# Acá TENEMOS QUE INDICAR la columna "race_id" al ultimo porque SI RE LANZAMOS EL NOTEBOOK y la columna no la colocamos al ultimo, 
# se generara un ERROR, dado la columna "race_id" se creó al ultimo en la tabla "f1_processed.results" la primera vez que se 
# ejecutó este notebook. Por tanto, colocarla en segundo lugar, solo generaría un desentendimiento en SPARK, generandose un ERROR.
results_final_df = results_final_df.select("result_id","driver_id","constructor_id","number","grid","position","position_text",
                                           "position_order","points","laps","time","milliseconds","fastest_lap","rank","fastest_lap_time",
                                           "fastest_lap_speed","data_source","file_date","ingestion_date","race_id")

In [ ]:
if (spark._jsparkSession.catalog().tableExists("f1_processed.results")):
  results_final_df.write.mode('overwrite').insertInto("f1_processed.results")
else:
  results_final_df.write.partitionBy('race_id').format('parquet').saveAsTable("f1_processed.results")

In [ ]:
spark.read.parquet("/mnt/formula1dl/processed/results").show(truncate=False)

+---------+---------+--------------+------+----+--------+-------------+--------------+------+----+----------+------------+-----------+----+----------------+-----------------+-----------+----------+-----------------------+-------+
|result_id|driver_id|constructor_id|number|grid|position|position_text|position_order|points|laps|time      |milliseconds|fastest_lap|rank|fastest_lap_time|fastest_lap_speed|data_source|file_date |ingestion_date         |race_id|
+---------+---------+--------------+------+----+--------+-------------+--------------+------+----+----------+------------+-----------+----+----------------+-----------------+-----------+----------+-----------------------+-------+
|19232    |657      |113           |14    |19  |1       |1            |1             |8.0   |200 |3:49:17.27|13757270    |null       |null|\N              |null             |Ergast     |2021-03-21|2023-06-15 16:29:32.103|800    |
|19233    |525      |114           |9     |3   |2       |2            |2        

### **Apuntes #3** - Carga Incremental para el `Método 3` (Parte 2)

Esta función nos genera una lista con el orden necesario de las columnas, dejando al final la columna **race_id**

1. **input_df** es el parámetro que ingresamos que corresponde al dataframe
2. **partition_column** es el parámetro que debemos ingresar igual a **race_id**
3. **input_df.schema.names** nos devuelve una lista con los nombres de las columnas del dataframe **input_df**

In [ ]:
# La función "re_arrange_partition_column()" se encuentra en el notebook "common_functions"
# Recordar que esta función se llamará al utilizar %run "../includes/common_functions"
def re_arrange_partition_column(input_df, partition_column):
  column_list = []
  for column_name in input_df.schema.names:
    if column_name != partition_column: # Si "column_name" es distinto a "race_id"
      column_list.append(column_name)  # Lo añadirá a la lista "column_list"
  column_list.append(partition_column) # Despues de ejecutarse el FOR, agrega a la lista "column_list" el valor de "race_id" al último
  output_df = input_df.select(column_list)
  return output_df

In [ ]:
re_arrange_partition_column(results_final_df, 'race_id')

Out[24]: DataFrame[result_id: int, driver_id: int, constructor_id: int, number: int, grid: int, position: int, position_text: string, position_order: int, points: float, laps: int, time: string, milliseconds: int, fastest_lap: int, rank: int, fastest_lap_time: string, fastest_lap_speed: float, data_source: string, file_date: string, ingestion_date: timestamp, race_id: int]

1. Esta función llama a la función **re_arrange_partition_column** de la cual obtiene el dataframe con el orden que necesitamos de las columnas, con la columna particionada **race_id** al último.
2. Finalmente se ejecuta la condición **IF** y si existiese la tabla, se sobrescribirán las particiones coincidentes
3. En caso contrario, se creará la tabla

In [ ]:
# La función "overwrite_partition()" se encuentra en el notebook "common_functions"
# Recordar que esta función se llamará al utilizar %run "../includes/common_functions"
def overwrite_partition(input_df, db_name, table_name, partition_column):
  output_df = re_arrange_partition_column(input_df, partition_column)
  spark.conf.set("spark.sql.sources.partitionOverwriteMode","dynamic")
  if (spark._jsparkSession.catalog().tableExists(f"{db_name}.{table_name}")):
    output_df.write.mode("overwrite").insertInto(f"{db_name}.{table_name}")
  else:
    output_df.write.mode("overwrite").partitionBy(partition_column).format("parquet").saveAsTable(f"{db_name}.{table_name}")

Por último, llamamos a la función:

In [ ]:
overwrite_partition(results_final_df, 'f1_processed', 'results', 'race_id')

### **Apuntes #4** - Carga Incremental para las tarnsformaciones

#### `race_results`

|En la transformacion incremental vamos a utilizar `race_id`, esta columna no la utilizamos en la transformacion para la carga completa. La columna `file_date` tampoco la utilizamos|
|---|

- results_df
    - results_race_id
    - result_file_date

- race_circuits_df
    - race_id
    - file_date

1. Debemos utilizar la columna **race_id**, pero esta se repite 2 veces en el dataframe **race_results_df**, una proveniente de **results_df** (`results_race_id`) y otra proveniente de **race_circuits_df** (`race_id`). Es por este motivo que cambiamos su nombre a **results_race_id** en el dataframe de **results_df**. De esta manera, no tendremos conflictos de nombre de columna (ambiguedad)
2. Hacemos lo mismo con la columna `file_date` del dataframe **results_df**, lo renombramos como `result_file_date`. Y nosotros vamos a utilizar `file_date`

In [ ]:
final_df = race_results_df.select("race_id","race_year", "race_name", "race_date", "circuit_location", "driver_name", "driver_number", "driver_nationality",
                                  "team", "grid", "fastest_lap", "race_time", "points", "position", "file_date") \
                          .withColumn("created_date",current_timestamp())

Dado que vamos a utilizar `race_id` nos permitirá utilizar la función `re_arrange_partition_column()` que nos ordenará las columnas de forma correcta dejando la columna `race_id` al ultimo. Y desde la función `overwrite_partition()` quien llama a `re_arrange_partition_column()` gestionará la creación de una tabla (directorio con particiones). Y las particiones son las que nos dan la posibilidad de trabajar de manera incremental, dado que gestionanando su escritura y sobrescritura podemos ir agregando nuevos registros sin afectar a los demas.

Despues se utiliza la funcion `overwrite_partition()` que exta explicada en **Apuntes #3** - Carga Incremental para el `Método 3` (Parte 2)

#### `driver_standings`

1. Este notebook se ejecutó para el directorio **2021-03-21**. 
2. Podemos reutilizarlo para los dos directorios restantes: **2021-03-28** y **2021-04-18** o para los que lleguen en el futuro
3. Solamente debemos modificar el parámetro del notebook **p_file_date**

In [ ]:
dbutils.widgets.text("p_file_date", "2021-03-21")
v_file_date = dbutils.widgets.get("p_file_date")

In [ ]:
%run "../includes/configuration"

In [ ]:
%run "../includes/common_functions"

##### Encontrar el **race_year** para los que se desea reprocesar los datos

In [ ]:
# El parámetro "presentation_folder_path" se encuentra en el notebook "configuration"
race_results_list = spark.read.parquet(f"{presentation_folder_path}/race_results") \
.filter(f"file_date = '{v_file_date}'") \
.select("race_year") \
.distinct() \
.collect()

In [ ]:
race_results_list

Out[23]: [Row(race_year=2018),
 Row(race_year=2013),
 Row(race_year=2014),
 Row(race_year=1989),
 Row(race_year=2020),
 Row(race_year=2012),
 Row(race_year=2016),
 Row(race_year=2010),
 Row(race_year=2011),
 Row(race_year=2007),
 Row(race_year=2015),
 Row(race_year=2019),
 Row(race_year=1954),
 Row(race_year=2017),
 Row(race_year=1990),
 Row(race_year=2006),
 Row(race_year=2008),
 Row(race_year=1977),
 Row(race_year=1988),
 Row(race_year=1991),
 Row(race_year=2009),
 Row(race_year=1974),
 Row(race_year=1978),
 Row(race_year=1953),
 Row(race_year=2005),
 Row(race_year=1981),
 Row(race_year=1975),
 Row(race_year=1961),
 Row(race_year=1994),
 Row(race_year=1973),
 Row(race_year=1979),
 Row(race_year=2004),
 Row(race_year=1982),
 Row(race_year=1965),
 Row(race_year=1972),
 Row(race_year=1992),
 Row(race_year=1976),
 Row(race_year=1952),
 Row(race_year=1971),
 Row(race_year=1960),
 Row(race_year=1987),
 Row(race_year=1980),
 Row(race_year=1983),
 Row(race_year=1963),
 Row(race_year=1984),
 

In [ ]:
race_year_list = []
for race_year in race_results_list:
  print(race_year)

Row(race_year=2018)
Row(race_year=2013)
Row(race_year=2014)
Row(race_year=1989)
Row(race_year=2020)
Row(race_year=2012)
Row(race_year=2016)
Row(race_year=2010)
Row(race_year=2011)
Row(race_year=2007)
Row(race_year=2015)
Row(race_year=2019)
Row(race_year=1954)
Row(race_year=2017)
Row(race_year=1990)
Row(race_year=2006)
Row(race_year=2008)
Row(race_year=1977)
Row(race_year=1988)
Row(race_year=1991)
Row(race_year=2009)
Row(race_year=1974)
Row(race_year=1978)
Row(race_year=1953)
Row(race_year=2005)
Row(race_year=1981)
Row(race_year=1975)
Row(race_year=1961)
Row(race_year=1994)
Row(race_year=1973)
Row(race_year=1979)
Row(race_year=2004)
Row(race_year=1982)
Row(race_year=1965)
Row(race_year=1972)
Row(race_year=1992)
Row(race_year=1976)
Row(race_year=1952)
Row(race_year=1971)
Row(race_year=1960)
Row(race_year=1987)
Row(race_year=1980)
Row(race_year=1983)
Row(race_year=1963)
Row(race_year=1984)
Row(race_year=1962)
Row(race_year=1959)
Row(race_year=2003)
Row(race_year=1956)
Row(race_year=1997)


In [ ]:
race_year_list = []
for race_year in race_results_list:
  race_year_list.append(race_year.race_year)
  
print(race_year_list)

[2018, 2013, 2014, 1989, 2020, 2012, 2016, 2010, 2011, 2007, 2015, 2019, 1954, 2017, 1990, 2006, 2008, 1977, 1988, 1991, 2009, 1974, 1978, 1953, 2005, 1981, 1975, 1961, 1994, 1973, 1979, 2004, 1982, 1965, 1972, 1992, 1976, 1952, 1971, 1960, 1987, 1980, 1983, 1963, 1984, 1962, 1959, 2003, 1956, 1997, 1950, 1985, 1970, 2001, 1993, 1955, 1995, 2000, 1986, 2002, 1951, 1996, 1998, 1958, 1999, 1957, 1968, 1967, 1964, 1966, 1969]


Esto lo simplico con la funcion: `df_column_to_list()`

In [ ]:
# Ejemplo
from pyspark.sql import SparkSession
from pyspark.sql import Row

# Crear una sesión de Spark
spark = SparkSession.builder \
    .appName("Ejemplo DataFrame") \
    .getOrCreate()

# Crear una lista de años
years = [2000 + i for i in range(11)]  # Genera años del 2000 al 2010

# Crear una lista de filas
rows = [Row(race_year=year) for year in years]

# Crear el DataFrame
df = spark.createDataFrame(rows)

# Mostrar el DataFrame
df.show()

```text
+---------+
|race_year|
+---------+
|     2000|
|     2001|
|     2002|
|     2003|
|     2004|
|     2005|
|     2006|
|     2007|
|     2008|
|     2009|
|     2010|
+---------+
```

In [ ]:
def df_column_to_list(input_df, column_name):
  df_row_list = input_df.select(column_name) \
                        .distinct() \
                        .collect()
  
  column_value_list = [row[column_name] for row in df_row_list]
  return column_value_list

In [ ]:
#                      row["race_year"] for Row(race_year=XXXX) in [Row(race_year=2018), Row(race_year=2013), Row(race_year=2014),Row(race_year=1989), ....]
# column_value_list = [row[column_name] for row in df_row_list]

df_column_to_list(df,"race_year")

```text
[2000, 2001, 2003, 2002, 2004, 2005, 2007, 2006, 2008, 2009, 2010]
```

Por lo tanto, es eso lo que hago aqui:

In [ ]:
race_list_year = df_column_to_list(race_results_list, "race_year")

In [ ]:
# El parámetro "presentation_folder_path" se encuentra en el notebook "configuration"
#race_results_df = spark.read.parquet(f"{presentation_folder_path}/race_results")
# Cuando filtremos por una fecha (por ejemplo "2021-03-21") va a buscar en las fechas de las carreras EL AÑO en el cual estan se
# realizaron y de acuerdo a ese año o esos años, se va a filtrar. Para el caso que filtremos por la fecha "2021-03-28", dicho
# directorio solo posee carreras que ocurrieron solo ese dia, por tanto, el dataframe se filtrará por el año "2021"
from pyspark.sql.functions import col

race_results_df = spark.read.parquet("/mnt/formula1dl/presentation/race_results") \
.filter(col("race_year").isin(race_year_list))

In [ ]:
race_results_df.show(truncate=False)

+---------+--------------------+-------------------+----------------+------------------+-------------+------------------+-----------+----+-----------+---------+------+--------+----------+-----------------------+-------+
|race_year|race_name           |race_date          |circuit_location|driver_name       |driver_number|driver_nationality|team       |grid|fastest_lap|race_time|points|position|file_date |created_date           |race_id|
+---------+--------------------+-------------------+----------------+------------------+-------------+------------------+-----------+----+-----------+---------+------+--------+----------+-----------------------+-------+
|2012     |Singapore Grand Prix|2012-09-23 12:00:00|Marina Bay      |Lewis Hamilton    |44           |British           |McLaren    |1   |14         |\N       |0.0   |null    |2021-03-21|2023-06-15 23:29:54.239|873    |
|2012     |Singapore Grand Prix|2012-09-23 12:00:00|Marina Bay      |Narain Karthikeyan|null         |Indian            

In [ ]:
from pyspark.sql.functions import sum, when, count, col

driver_standings_df = race_results_df \
.groupBy("race_year", "driver_name", "driver_nationality", "team") \
.agg(sum("points").alias("total_points"),
     count(when(col("position") == 1, True)).alias("wins")).sort(col('wins').desc()) 
# Cuando posicion es igual a 1, entonces, es igual a 1 (True = 1 y False = 0)

In [ ]:
driver_standings_df.show(truncate=False)

+---------+------------------+------------------+--------+------------+----+
|race_year|driver_name       |driver_nationality|team    |total_points|wins|
+---------+------------------+------------------+--------+------------+----+
|2013     |Sebastian Vettel  |German            |Red Bull|397.0       |13  |
|2004     |Michael Schumacher|German            |Ferrari |148.0       |13  |
|2014     |Lewis Hamilton    |British           |Mercedes|384.0       |11  |
|2020     |Lewis Hamilton    |British           |Mercedes|347.0       |11  |
|2011     |Sebastian Vettel  |German            |Red Bull|392.0       |11  |
|2018     |Lewis Hamilton    |British           |Mercedes|408.0       |11  |
|2019     |Lewis Hamilton    |British           |Mercedes|413.0       |11  |
|2002     |Michael Schumacher|German            |Ferrari |144.0       |11  |
|2016     |Lewis Hamilton    |British           |Mercedes|380.0       |10  |
|2015     |Lewis Hamilton    |British           |Mercedes|381.0       |10  |

In [ ]:
driver_standings_df.filter("race_year = 2020").show(truncate=False)

+---------+------------------+------------------+------------+------------+----+
|race_year|driver_name       |driver_nationality|team        |total_points|wins|
+---------+------------------+------------------+------------+------------+----+
|2020     |Lewis Hamilton    |British           |Mercedes    |347.0       |11  |
|2020     |Valtteri Bottas   |Finnish           |Mercedes    |223.0       |2   |
|2020     |Max Verstappen    |Dutch             |Red Bull    |214.0       |2   |
|2020     |Pierre Gasly      |French            |AlphaTauri  |75.0        |1   |
|2020     |Sergio Pérez      |Mexican           |Racing Point|125.0       |1   |
|2020     |Lance Stroll      |Canadian          |Racing Point|75.0        |0   |
|2020     |Kevin Magnussen   |Danish            |Haas F1 Team|1.0         |0   |
|2020     |Antonio Giovinazzi|Italian           |Alfa Romeo  |4.0         |0   |
|2020     |Carlos Sainz      |Spanish           |McLaren     |105.0       |0   |
|2020     |Nicholas Latifi  

In [ ]:
from pyspark.sql.window import Window
from pyspark.sql.functions import desc, rank, asc

driver_rank_spec = Window.partitionBy("race_year").orderBy(desc("total_points"), desc("wins"))
final_df = driver_standings_df.withColumn("rank", rank().over(driver_rank_spec))

In [ ]:
final_df.show(truncate=False)

+---------+------------------+------------------+------------+------------+----+----+
|race_year|driver_name       |driver_nationality|team        |total_points|wins|rank|
+---------+------------------+------------------+------------+------------+----+----+
|1950     |Nino Farina       |Italian           |Alfa Romeo  |30.0        |3   |1   |
|1950     |Luigi Fagioli     |Italian           |Alfa Romeo  |28.0        |0   |2   |
|1950     |Juan Fangio       |Argentine         |Alfa Romeo  |27.0        |3   |3   |
|1950     |Louis Rosier      |French            |Talbot-Lago |13.0        |0   |4   |
|1950     |Alberto Ascari    |Italian           |Ferrari     |11.0        |0   |5   |
|1950     |Johnnie Parsons   |American          |Kurtis Kraft|9.0         |1   |6   |
|1950     |Bill Holland      |American          |Deidt       |6.0         |0   |7   |
|1950     |Prince Bira       |Thai              |Maserati    |5.0         |0   |8   |
|1950     |Mauri Rose        |American          |Deidt

In [ ]:
final_df.filter("race_year = 2020").show(truncate=False)

+---------+------------------+------------------+------------+------------+----+----+
|race_year|driver_name       |driver_nationality|team        |total_points|wins|rank|
+---------+------------------+------------------+------------+------------+----+----+
|2020     |Lewis Hamilton    |British           |Mercedes    |347.0       |11  |1   |
|2020     |Valtteri Bottas   |Finnish           |Mercedes    |223.0       |2   |2   |
|2020     |Max Verstappen    |Dutch             |Red Bull    |214.0       |2   |3   |
|2020     |Sergio Pérez      |Mexican           |Racing Point|125.0       |1   |4   |
|2020     |Daniel Ricciardo  |Australian        |Renault     |119.0       |0   |5   |
|2020     |Carlos Sainz      |Spanish           |McLaren     |105.0       |0   |6   |
|2020     |Alexander Albon   |Thai              |Red Bull    |105.0       |0   |6   |
|2020     |Charles Leclerc   |Monegasque        |Ferrari     |98.0        |0   |8   |
|2020     |Lando Norris      |British           |McLar

##### Escribir datos en el contenedor **presentation** del ADLS como **parquet** y crear la tabla **driver_standings** en la base de datos **f1_presentation**

In [ ]:
dbutils.fs.rm("/mnt/formula1dl/presentation/driver_standings",True)

Out[34]: True

##### **Vamos a analizar las funciones**

Esta función nos genera una lista con el orden necesario de las columnas, dejando al final la columna **race_id**

1. **input_df** es el parámetro que ingresamos que corresponde al dataframe
2. **partition_column** es el parámetro que debemos ingresar igual a **race_id**
3. **input_df.schema.names** nos devuelve una lista con los nombres de las columnas del dataframe **input_df**

In [ ]:
# La función "re_arrange_partition_column()" se encuentra en el notebook "common_functions"
# Recordar que esta función se llamará al utilizar %run "../includes/common_functions"
def re_arrange_partition_column(input_df, partition_column):
  column_list = []
  for column_name in input_df.schema.names:
    if column_name != partition_column: # Si "column_name" es distinto a "race_year"
      column_list.append(column_name)  # Lo añadirá a la lista "column_list"
  column_list.append(partition_column) # Despues de ejecutarse el FOR, agrega a la lista "column_list" el valor de "race_year" al último
  output_df = input_df.select(column_list)
  return output_df

In [ ]:
re_arrange_partition_column(final_df, 'race_year')

Out[36]: DataFrame[driver_name: string, driver_nationality: string, team: string, total_points: double, wins: bigint, rank: int, race_year: int]

1. Esta función llama a la función **re_arrange_partition_column** de la cual obtiene el dataframe con el orden que necesitamos de las columnas, con la columna particionada **race_year** al último.
2. Finalmente se ejecuta la condición **IF** y si existiese la tabla, se sobrescribirán las particiones coincidentes
3. En caso contrario, se creará la tabla

In [ ]:
# La función "overwrite_partition()" se encuentra en el notebook "common_functions"
# Recordar que esta función se llamará al utilizar %run "../includes/common_functions"
def overwrite_partition(input_df, db_name, table_name, partition_column):
  output_df = re_arrange_partition_column(input_df, partition_column)
  spark.conf.set("spark.sql.sources.partitionOverwriteMode","dynamic")
  if (spark._jsparkSession.catalog().tableExists(f"{db_name}.{table_name}")):
    output_df.write.mode("overwrite").insertInto(f"{db_name}.{table_name}")
  else:
    output_df.write.mode("overwrite").partitionBy(partition_column).format("parquet").saveAsTable(f"{db_name}.{table_name}")

Por último, llamamos a la función:

In [ ]:
overwrite_partition(final_df, 'f1_presentation', 'driver_standings', 'race_year')

In [ ]:
spark.read.parquet("/mnt/formula1dl/presentation/driver_standings").show(truncate=False)

+---------------------+------------------+------------+------------+----+----+---------+
|driver_name          |driver_nationality|team        |total_points|wins|rank|race_year|
+---------------------+------------------+------------+------------+----+----+---------+
|Alberto Ascari       |Italian           |Ferrari     |53.5        |6   |1   |1952     |
|Nino Farina          |Italian           |Ferrari     |27.0        |0   |2   |1952     |
|Piero Taruffi        |Italian           |Ferrari     |22.0        |1   |3   |1952     |
|Rudi Fischer         |Swiss             |Ferrari     |10.0        |0   |4   |1952     |
|Mike Hawthorn        |British           |Cooper      |10.0        |0   |4   |1952     |
|Robert Manzon        |French            |Gordini     |9.0         |0   |6   |1952     |
|Troy Ruttman         |American          |Kuzma       |8.0         |1   |7   |1952     |
|Luigi Villoresi      |Italian           |Ferrari     |8.0         |0   |8   |1952     |
|José Froilán Gonzále

In [ ]:
%sql
SELECT * FROM f1_presentation.driver_standings;

driver_name driver_nationality team total_points wins rank race_year Alberto Ascari Italian Ferrari 53.5 6 1 1952 Nino Farina Italian Ferrari 27.0 0 2 1952 Piero Taruffi Italian Ferrari 22.0 1 3 1952 Rudi Fischer Swiss Ferrari 10.0 0 4 1952 Mike Hawthorn British Cooper 10.0 0 4 1952 Robert Manzon French Gordini 9.0 0 6 1952 Troy Ruttman American Kuzma 8.0 1 7 1952 Luigi Villoresi Italian Ferrari 8.0 0 8 1952 José Froilán González Argentine Maserati 6.5 0 9 1952 Jean Behra French Gordini 6.0 0 10 1952 Jim Rathmann American Kurtis Kraft 6.0 0 10 1952 Sam Hanks American Kurtis Kraft 4.0 0 12 1952 Dennis Poore British Connaught 3.0 0 13 1952 Duane Carter American Lesovsky 3.0 0 13 1952 Ken Wharton British Frazer Nash 3.0 0 13 1952 Felice Bonetto Italian Maserati 2.0 0 16 1952 Alan Brown British Cooper 2.0 0 16 1952 Art Cross American Kurtis Kraft 2.0 0 16 1952 Eric Thompson British Connaught 2.0 0 16 1952 Paul Frère Belgian HWM 2.0 0 16 1952 Maurice Trintignant French Simca 2.0 0 16 1952 Bill Vukovich American Kurtis Kraft 1.0 0 22 1952 Toulo de Graffenried Swiss Maserati 0.0 0 23 1952 André Simon French Ferrari 0.0 0 23 1952 Eric Brandon British Cooper 0.0 0 23 1952 Johnny Claes Belgian Simca 0.0 0 23 1952 Piero Dusio Italian Cisitalia 0.0 0 23 1952 Franco Rol Italian Maserati 0.0 0 23 1952 Ken Wharton British Cooper 0.0 0 23 1952 Gino Bianco Brazilian Maserati 0.0 0 23 1952 Charles de Tornaco Belgian Ferrari 0.0 0 23 1952 Lance Macklin British HWM 0.0 0 23 1952 Hans von Stuck German Ferrari 0.0 0 23 1952 Peter Whitehead British Ferrari 0.0 0 23 1952 Alberto Crespo Argentine Maserati 0.0 0 23 1952 Stirling Moss British Connaught 0.0 0 23 1952 Tony Gaze Australian HWM 0.0 0 23 1952 Chico Landi Brazilian Maserati 0.0 0 23 1952 Peter Collins British HWM 0.0 0 23 1952 Louis Rosier French Ferrari 0.0 0 23 1952 Eitel Cantoni Uruguayan Maserati 0.0 0 23 1952 Maurice Trintignant French Gordini 0.0 0 23 1952 Kenneth McAlpine British Connaught 0.0 0 23 1952 Élie Bayol French OSCA 0.0 0 23 1952 Bill Aston British Aston Butterworth 0.0 0 23 1952 Andy Linden American Kurtis Kraft 0.0 0 23 1952 Johnnie Parsons American Kurtis Kraft 0.0 0 23 1952 Joe James American Kurtis Kraft 0.0 0 23 1952 Fred Agabashian American Kurtis Kraft 0.0 0 23 1952 Duke Nalon American Kurtis Kraft 0.0 0 23 1952 Gene Hartley American Kurtis Kraft 0.0 0 23 1952 Bill Schindler American Stevens 0.0 0 23 1952 Bob Sweikert American Kurtis Kraft 0.0 0 23 1952 Jack McGrath American Kurtis Kraft 0.0 0 23 1952 Jimmy Reece American Kurtis Kraft 0.0 0 23 1952 Chet Miller American Kurtis Kraft 0.0 0 23 1952 Tony Bettenhausen American Deidt 0.0 0 23 1952 George Connor American Kurtis Kraft 0.0 0 23 1952 Chuck Stevenson American Kurtis Kraft 0.0 0 23 1952 Cliff Griffith American Kurtis Kraft 0.0 0 23 1952 Manny Ayulo American Lesovsky 0.0 0 23 1952 Rodger Ward American Kurtis Kraft 0.0 0 23 1952 Bob Scott American Kurtis Kraft 0.0 0 23 1952 Eddie Johnson American Trevis 0.0 0 23 1952 George Fonder American Sherman 0.0 0 23 1952 Travis Webb American Bromme 0.0 0 23 1952 Bobby Ball American Stevens 0.0 0 23 1952 Henry Banks American Lesovsky 0.0 0 23 1952 Jimmy Bryan American Kurtis Kraft 0.0 0 23 1952 Jim Rigsby American Watson 0.0 0 23 1952 Johnny McDowell American Kurtis Kraft 0.0 0 23 1952 Stirling Moss British ERA 0.0 0 23 1952 Peter Hirt Swiss Ferrari 0.0 0 23 1952 Harry Schell American Maserati 0.0 0 23 1952 Roy Salvadori British Ferrari 0.0 0 23 1952 Tony Crook British Frazer Nash 0.0 0 23 1952 Graham Whitehead British Alta 0.0 0 23 1952 Prince Bira Thai Gordini 0.0 0 23 1952 Ken Downing British Connaught 0.0 0 23 1952 Duncan Hamilton British HWM 0.0 0 23 1952 Reg Parnell British Cooper 0.0 0 23 1952 David Murray British Cooper 0.0 0 23 1952 Rudolf Krause East German BMW 0.0 0 23 1952 Ernst Klodwig East German BMW 0.0 0 23 1952 Marcel Balsa French BMW 0.0 0 23 1952 Piero Carini Italian Ferrari 0.0 0 23 1952 Josef Peters German Veritas 0.0 0 23 1952 Helmut Niedermayr German AF

#### `constructor_standings`

1. Este notebook se ejecutó para el directorio **2021-03-21**. 
2. Podemos reutilizarlo para los dos directorios restantes: **2021-03-28** y **2021-04-18** o para los que lleguen en el futuro
3. Solamente debemos modificar el parámetro del notebook **p_file_date**

In [ ]:
dbutils.widgets.text("p_file_date", "2021-03-21")
v_file_date = dbutils.widgets.get("p_file_date")

In [ ]:
v_file_date

Out[24]: '2021-03-21'

In [ ]:
%run "../includes/configuration"

In [ ]:
%run "../includes/common_functions"

#### Encontrar el **race_year** para los que se deben reprocesar los datos

In [ ]:
# El parámetro "presentation_folder_path" se encuentra en el notebook "configuration"
race_results_df = spark.read.parquet(f"{presentation_folder_path}/race_results")
.filter(f"file_date = '{v_file_date}'")

In [ ]:
race_results_df.show(truncate=False)

+---------+--------------------+-------------------+----------------+------------------+-------------+------------------+-----------+----+-----------+---------+------+--------+----------+-----------------------+-------+
|race_year|race_name           |race_date          |circuit_location|driver_name       |driver_number|driver_nationality|team       |grid|fastest_lap|race_time|points|position|file_date |created_date           |race_id|
+---------+--------------------+-------------------+----------------+------------------+-------------+------------------+-----------+----+-----------+---------+------+--------+----------+-----------------------+-------+
|2012     |Singapore Grand Prix|2012-09-23 12:00:00|Marina Bay      |Lewis Hamilton    |44           |British           |McLaren    |1   |14         |\N       |0.0   |null    |2021-03-21|2023-06-15 23:29:54.239|873    |
|2012     |Singapore Grand Prix|2012-09-23 12:00:00|Marina Bay      |Narain Karthikeyan|null         |Indian            

In [ ]:
# La función "df_column_to_list()" se encuentra en el notebook "common_functions"
# Recordar que esta función se llamará al utilizar %run "../includes/common_functions"
def df_column_to_list(input_df, column_name):
  df_row_list = input_df.select(column_name) \
                        .distinct() \
                        .collect()
  
  column_value_list = [row[column_name] for row in df_row_list]
  return column_value_list

In [ ]:
race_year_list = df_column_to_list(race_results_df, 'race_year')

In [ ]:
race_year_list

Out[50]: [2018,
 2013,
 2014,
 1989,
 2020,
 2012,
 2016,
 2010,
 2011,
 2007,
 2015,
 2019,
 1954,
 2017,
 1990,
 2006,
 2008,
 1977,
 1988,
 1991,
 2009,
 1974,
 1978,
 1953,
 2005,
 1981,
 1975,
 1961,
 1994,
 1973,
 1979,
 2004,
 1982,
 1965,
 1972,
 1992,
 1976,
 1952,
 1971,
 1960,
 1987,
 1980,
 1983,
 1963,
 1984,
 1962,
 1959,
 2003,
 1956,
 1997,
 1950,
 1985,
 1970,
 2001,
 1993,
 1955,
 1995,
 2000,
 1986,
 2002,
 1951,
 1996,
 1998,
 1958,
 1999,
 1957,
 1968,
 1967,
 1964,
 1966,
 1969]

In [ ]:
# El parámetro "presentation_folder_path" se encuentra en el notebook "configuration"
from pyspark.sql.functions import col

race_results_df = spark.read.parquet(f"{presentation_folder_path}/race_results") \
.filter(col("race_year").isin(race_year_list))

In [ ]:
from pyspark.sql.functions import sum, when, count, col

constructor_standings_df = race_results_df \
.groupBy("race_year", "team") \
.agg(sum("points").alias("total_points"),
     count(when(col("position") == 1, True)).alias("wins")).sort(col('wins').desc()) 
# Cuando posicion es igual a 1, entonces, es igual a 1 (True = 1 y False = 0)

In [ ]:
constructor_standings_df.show(truncate=False)

+---------+--------+------------+----+
|race_year|team    |total_points|wins|
+---------+--------+------------+----+
|2016     |Mercedes|765.0       |19  |
|2014     |Mercedes|701.0       |16  |
|2015     |Mercedes|703.0       |16  |
|1988     |McLaren |199.0       |15  |
|2019     |Mercedes|739.0       |15  |
|2004     |Ferrari |262.0       |15  |
|2002     |Ferrari |221.0       |15  |
|2013     |Red Bull|596.0       |13  |
|2020     |Mercedes|573.0       |13  |
|2011     |Red Bull|650.0       |12  |
|2017     |Mercedes|668.0       |12  |
|1984     |McLaren |143.5       |12  |
|1996     |Williams|175.0       |12  |
|2018     |Mercedes|655.0       |11  |
|1995     |Benetton|147.0       |11  |
|1989     |McLaren |141.0       |10  |
|2005     |McLaren |182.0       |10  |
|1992     |Williams|164.0       |10  |
|1993     |Williams|168.0       |10  |
|2000     |Ferrari |170.0       |10  |
+---------+--------+------------+----+
only showing top 20 rows



In [ ]:
constructor_standings_df.filter("race_year = 2020").show(truncate=False)

+---------+------------+------------+----+
|race_year|team        |total_points|wins|
+---------+------------+------------+----+
|2020     |Mercedes    |573.0       |13  |
|2020     |Red Bull    |319.0       |2   |
|2020     |AlphaTauri  |107.0       |1   |
|2020     |Racing Point|210.0       |1   |
|2020     |Haas F1 Team|3.0         |0   |
|2020     |McLaren     |202.0       |0   |
|2020     |Ferrari     |131.0       |0   |
|2020     |Williams    |0.0         |0   |
|2020     |Alfa Romeo  |8.0         |0   |
|2020     |Renault     |181.0       |0   |
+---------+------------+------------+----+



In [ ]:
from pyspark.sql.window import Window
from pyspark.sql.functions import desc, rank, asc

constructor_rank_spec = Window.partitionBy("race_year").orderBy(desc("total_points"), desc("wins"))
final_df = constructor_standings_df.withColumn("rank", rank().over(constructor_rank_spec))

In [ ]:
final_df.show(truncate=False)

+---------+------------+------------+----+----+
|race_year|team        |total_points|wins|rank|
+---------+------------+------------+----+----+
|1950     |Alfa Romeo  |89.0        |6   |1   |
|1950     |Ferrari     |21.0        |0   |2   |
|1950     |Talbot-Lago |20.0        |0   |3   |
|1950     |Kurtis Kraft|14.0        |1   |4   |
|1950     |Maserati    |11.0        |0   |5   |
|1950     |Deidt       |10.0        |0   |6   |
|1950     |Simca       |3.0         |0   |7   |
|1950     |Rae         |0.0         |0   |8   |
|1950     |Langley     |0.0         |0   |8   |
|1950     |Lesovsky    |0.0         |0   |8   |
|1950     |Ewing       |0.0         |0   |8   |
|1950     |Stevens     |0.0         |0   |8   |
|1950     |Marchese    |0.0         |0   |8   |
|1950     |Nichels     |0.0         |0   |8   |
|1950     |Wetteroth   |0.0         |0   |8   |
|1950     |Watson      |0.0         |0   |8   |
|1950     |Olson       |0.0         |0   |8   |
|1950     |Adams       |0.0         |0  

In [ ]:
final_df.filter("race_year = 2020").show(truncate=False)

+---------+------------+------------+----+----+
|race_year|team        |total_points|wins|rank|
+---------+------------+------------+----+----+
|2020     |Mercedes    |573.0       |13  |1   |
|2020     |Red Bull    |319.0       |2   |2   |
|2020     |Racing Point|210.0       |1   |3   |
|2020     |McLaren     |202.0       |0   |4   |
|2020     |Renault     |181.0       |0   |5   |
|2020     |Ferrari     |131.0       |0   |6   |
|2020     |AlphaTauri  |107.0       |1   |7   |
|2020     |Alfa Romeo  |8.0         |0   |8   |
|2020     |Haas F1 Team|3.0         |0   |9   |
|2020     |Williams    |0.0         |0   |10  |
+---------+------------+------------+----+----+



#### Escribir datos en el contenedor **presentation** del ADLS como **parquet** y crear la tabla **constructor_standings** en la base de datos **f1_presentation**

#### **Vamos a analizar las funciones**

Esta función nos genera una lista con el orden necesario de las columnas, dejando al final la columna **race_year**

1. **input_df** es el parámetro que ingresamos que corresponde al dataframe
2. **partition_column** es el parámetro que debemos ingresar igual a **race_year**
3. **input_df.schema.names** nos devuelve una lista con los nombres de las columnas del dataframe **input_df**

In [ ]:
# La función "re_arrange_partition_column()" se encuentra en el notebook "common_functions"
# Recordar que esta función se llamará al utilizar %run "../includes/common_functions"
def re_arrange_partition_column(input_df, partition_column):
  column_list = []
  for column_name in input_df.schema.names:
    if column_name != partition_column: # Si "column_name" es distinto a "race_year"
      column_list.append(column_name)  # Lo añadirá a la lista "column_list"
  column_list.append(partition_column) # Despues de ejecutarse el FOR, agrega a la lista "column_list" el valor de "race_year" al último
  output_df = input_df.select(column_list)
  return output_df

In [ ]:
re_arrange_partition_column(final_df, 'race_year')

Out[45]: DataFrame[team: string, total_points: double, wins: bigint, rank: int, race_year: int]

1. Esta función llama a la función **re_arrange_partition_column** de la cual obtiene el dataframe con el orden que necesitamos de las columnas, con la columna particionada **race_year** al último.
2. Finalmente se ejecuta la condición **IF** y si existiese la tabla, se sobrescribirán las particiones coincidentes
3. En caso contrario, se creará la tabla

In [ ]:
# La función "overwrite_partition()" se encuentra en el notebook "common_functions"
# Recordar que esta función se llamará al utilizar %run "../includes/common_functions"
def overwrite_partition(input_df, db_name, table_name, partition_column):
  output_df = re_arrange_partition_column(input_df, partition_column)
  spark.conf.set("spark.sql.sources.partitionOverwriteMode","dynamic")
  if (spark._jsparkSession.catalog().tableExists(f"{db_name}.{table_name}")):
    output_df.write.mode("overwrite").insertInto(f"{db_name}.{table_name}")
  else:
    output_df.write.mode("overwrite").partitionBy(partition_column).format("parquet").saveAsTable(f"{db_name}.{table_name}")

Por último, llamamos a la función:

In [ ]:
# La función "overwrite_partition()" se encuentra en el notebook "common_functions"
# Recordar que esta función se llamará al utilizar %run "../includes/common_functions"
overwrite_partition(final_df, 'f1_presentation', 'constructor_standings', 'race_year')

In [ ]:
spark.read.parquet("/mnt/formula1dl/presentation/constructor_standings").show(truncate=False)

+------------------+------------+----+----+---------+
|team              |total_points|wins|rank|race_year|
+------------------+------------+----+----+---------+
|Cooper-Climax     |102.0       |6   |1   |1960     |
|Team Lotus        |52.0        |2   |2   |1960     |
|Ferrari           |43.0        |1   |3   |1960     |
|Watson            |14.0        |1   |4   |1960     |
|BRM               |8.0         |0   |5   |1960     |
|Epperly           |4.0         |0   |6   |1960     |
|Phillips          |3.0         |0   |7   |1960     |
|Cooper-Maserati   |3.0         |0   |7   |1960     |
|Cooper-Castellotti|3.0         |0   |7   |1960     |
|Lesovsky          |2.0         |0   |10  |1960     |
|Trevis            |1.0         |0   |11  |1960     |
|Porsche           |1.0         |0   |11  |1960     |
|Meskowski         |0.0         |0   |13  |1960     |
|Ewing             |0.0         |0   |13  |1960     |
|Christensen       |0.0         |0   |13  |1960     |
|Kurtis Kraft      |0.0     

In [ ]:
%sql
SELECT * FROM f1_presentation.constructor_standings;

team total_points wins rank race_year Cooper-Climax 102.0 6 1 1960 Team Lotus 52.0 2 2 1960 Ferrari 43.0 1 3 1960 Watson 14.0 1 4 1960 BRM 8.0 0 5 1960 Epperly 4.0 0 6 1960 Phillips 3.0 0 7 1960 Cooper-Maserati 3.0 0 7 1960 Cooper-Castellotti 3.0 0 7 1960 Lesovsky 2.0 0 10 1960 Trevis 1.0 0 11 1960 Porsche 1.0 0 11 1960 Meskowski 0.0 0 13 1960 Ewing 0.0 0 13 1960 Christensen 0.0 0 13 1960 Kurtis Kraft 0.0 0 13 1960 Kuzma 0.0 0 13 1960 JBW 0.0 0 13 1960 Behra-Porsche 0.0 0 13 1960 Aston Martin 0.0 0 13 1960 Maserati 0.0 0 13 1960 Scarab 0.0 0 13 1960 Vanwall 0.0 0 13 1960 Cooper-Climax 97.5 5 1 1959 Ferrari 67.0 2 2 1959 BRM 22.5 1 3 1959 Watson 14.0 1 4 1959 Lesovsky 5.0 0 5 1959 Epperly 5.0 0 5 1959 Team Lotus 5.0 0 5 1959 Sutton 0.0 0 8 1959 Elder 0.0 0 8 1959 Kurtis Kraft 0.0 0 8 1959 Christensen 0.0 0 8 1959 Kuzma 0.0 0 8 1959 Phillips 0.0 0 8 1959 Dunn 0.0 0 8 1959 Moore 0.0 0 8 1959 Fry 0.0 0 8 1959 Maserati 0.0 0 8 1959 Cooper-Borgward 0.0 0 8 1959 Aston Martin 0.0 0 8 1959 Cooper-Maserati 0.0 0 8 1959 Vanwall 0.0 0 8 1959 JBW 0.0 0 8 1959 Porsche 0.0 0 8 1959 Tec-Mec 0.0 0 8 1959 Connaught 0.0 0 8 1959 Cooper-OSCA 0.0 0 8 1959 Ferrari 89.5 6 1 1975 McLaren 65.0 3 2 1975 Brabham 61.0 2 3 1975 Hesketh 33.0 1 4 1975 Tyrrell 32.0 1 5 1975 Shadow-Ford 9.5 0 6 1975 March 9.0 1 7 1975 Team Lotus 9.0 0 8 1975 Williams 6.0 0 9 1975 Parnelli 5.0 0 10 1975 Embassy Hill 3.0 0 11 1975 Penske 2.0 0 12 1975 Ensign 1.0 0 13 1975 Surtees 0.0 0 14 1975 Maki 0.0 0 14 1975 Fittipaldi 0.0 0 14 1975 BRM 0.0 0 14 1975 Shadow-Matra 0.0 0 14 1975 Lola 0.0 0 14 1975 Lyncar 0.0 0 14 1975 McLaren 141.0 10 1 1989 Williams 77.0 2 2 1989 Ferrari 59.0 3 3 1989 Benetton 39.0 1 4 1989 Tyrrell 16.0 0 5 1989 Team Lotus 15.0 0 6 1989 Arrows 13.0 0 7 1989 Brabham 8.0 0 8 1989 Dallara 8.0 0 8 1989 Onyx 6.0 0 10 1989 Minardi 6.0 0 10 1989 March 4.0 0 12 1989 Rial 3.0 0 13 1989 Ligier 3.0 0 13 1989 Larrousse 1.0 0 15 1989 AGS 1.0 0 15 1989 Coloni 0.0 0 17 1989 Euro Brun 0.0 0 17 1989 Zakspeed 0.0 0 17 1989 Osella 0.0 0 17 1989 Ferrari 120.5 7 1 1952 Gordini 15.0 0 2 1952 Kurtis Kraft 13.0 0 3 1952 Cooper 12.0 0 4 1952 Maserati 8.5 0 5 1952 Kuzma 8.0 1 6 1952 Connaught 5.0 0 7 1952 Lesovsky 3.0 0 8 1952 Frazer Nash 3.0 0 8 1952 Simca 2.0 0 10 1952 HWM 2.0 0 10 1952 OSCA 0.0 0 12 1952 Cisitalia 0.0 0 12 1952 Aston Butterworth 0.0 0 12 1952 Sherman 0.0 0 12 1952 Watson 0.0 0 12 1952 Deidt 0.0 0 12 1952 Bromme 0.0 0 12 1952 Stevens 0.0 0 12 1952 Trevis 0.0 0 12 1952 Alta 0.0 0 12 1952 ERA 0.0 0 12 1952 Veritas 0.0 0 12 1952 AFM 0.0 0 12 1952 BMW 0.0 0 12 1952 Ferrari 99.0 6 1 1976 McLaren 88.0 6 2 1976 Tyrrell 88.0 1 3 1976 Team Lotus 32.0 1 4 1976 Penske 20.0 1 5 1976 Ligier 20.0 0 6 1976 March 19.0 1 7 1976 Brabham-Alfa Romeo 11.0 0 8 1976 Shadow 10.0 0 9 1976 Surtees 7.0 0 10 1976 Fittipaldi 3.0 0 11 1976 Ensign 2.0 0 12 1976 Parnelli 1.0 0 13 1976 Hesketh 0.0 0 14 1976 Wolf 0.0 0 14 1976 Brabham-Ford 0.0 0 14 1976 Boro 0.0 0 14 1976 Williams 0.0 0 14 1976 Kojima 0.0 0 14 1976 Maki 0.0 0 14 1976 BRM 0.0 0 14 1976 Ferrari 90.0 3 1 1974 McLaren 87.0 4 2 1974 Tyrrell 59.0 2 3 1974 Team Lotus 47.0 3 4 1974 Brabham 46.0 3 5 1974 Hesketh 15.0 0 6 1974 BRM 10.0 0 7 1974 Shadow 7.0 0 8 1974 March 6.0 0 9 1974 Iso Marlboro 4.0 0 10 1974 Surtees 3.0 0 11 1974 Lola 1.0 0 12 1974 Token 0.0 0 13 1974 Maki 0.0 0 13 1974 Amon 0.0 0 13 1974 Trojan 0.0 0 13 1974 Ensign 0.0 0 13 1974 Lyncar 0.0 0 13 1974 Penske 0.0 0 13 1974 Parnelli 0.0 0 13 1974 Ferrari 113.0 6 1 1979 Williams 75.0 5 2 1979 Ligier 61.0 3 3 1979 Team Lotus 39.0 0 4 1979 Tyrrell 28.0 0 5 1979 Renault 26.0 1 6 1979 McLaren 15.0 0 7 1979 Brabham-Alfa Romeo 7.0 0 8 1979 Arrows 5.0 0 9 1979 Shadow 3.0 0 10 1979 ATS 2.0 0 11 1979 Fittipaldi 1.0 0 12 1979 Rebaque 0.0 0 13 1979 Brabham-Ford 0.0 0 13 1979 Ensign 0.0 0 13 1979 Merzario 0.0 0 13 1979 Alfa Romeo 0.0 0 13 1979 Wolf 0.0 0 13 1979 Kauhsen 0.0 0 13 1979 Ferrari 114.0 4 1 1977 Team Lotus 67.0 5 2 1977 McLaren 65.0 3 3 1977 Wolf 55.0 3 4 1977 Brabham 27.0 0 5 1